src/appleseed.qtcommon/widgets/chartwidget.cpp

```diff
@@ -59,51 +59,73 @@ namespace qtcommon {
 // ChartBase class implementation.
 //
 
-ChartBase::ChartBase()
-  : m_equidistant(false)
-  , m_margin(15.0, 15.0)
-  , m_grid_brush(QBrush(QColor(50, 50, 50, 255)))
+void ChartBase::set_equidistant(const bool equidistant)
 {
+    m_equidistant = equidistant;
 }
 
-void ChartBase::set_equidistant(const bool equidistant)
+void ChartBase::set_origin(const Vector2d& origin)
 {
-    m_equidistant = equidistant;
+    m_origin = origin;
 }
 
 void ChartBase::set_grid_brush(const QBrush& brush)
 {
     m_grid_brush = brush;
 }
 
-void ChartBase::set_tooltip_formatter(std::unique_ptr<IToolTipFormatter> formatter)
+void ChartBase::set_legend_brush(const QBrush& brush)
+{
+    m_legend_brush = brush;
+}
+
+void ChartBase::set_horizontal_legend_formatter(const ValueFormatter& formatter)
+{
+    m_horizontal_legend_formatter = formatter;
+}
+
+void ChartBase::set_vertical_legend_formatter(const ValueFormatter& formatter)
+{
+    m_vertical_legend_formatter = formatter;
+}
+
+void ChartBase::set_tooltip_formatter(const TooltipFormatter& formatter)
 {
-    m_tooltip_formatter = std::move(formatter);
+    m_tooltip_formatter = formatter;
 }
 
-void ChartBase::add_point(const Vector2d& p)
+size_t ChartBase::size() const
+{
+    return m_original_points.size();
+}
+
+void ChartBase::push_back(const Vector2d& p)
 {
     m_original_points.push_back(p);
 }
 
-void ChartBase::add_point(const double x, const double y)
+void ChartBase::push_back(const double x, const double y)
 {
     m_original_points.emplace_back(x, y);
 }
 
+void ChartBase::pop_front()
+{
+    m_original_points.pop_front();
+}
+
 void ChartBase::prepare_drawing(QPainter& painter)
 {
     m_points = m_original_points;
 
     if (m_equidistant)
     {
-        for (size_t i = 0; i < m_points.size(); ++i)
+        for (size_t i = 0, e = m_points.size(); i < e; ++i)
             m_points[i].x = static_cast<double>(i);
-
     }
 
     m_points_bbox = compute_points_bbox();
-    m_rcp_points_bbox_extent = Vector2d(1.0) / m_points_bbox.extent();
+    m_rcp_points_bbox_extent = m_points_bbox.is_valid() ? Vector2d(1.0) / m_points_bbox.extent() : Vector2d(1.0);   // todo: fix!
```

Comment:
Can it be fixed now ?

Reply:
I believe the code is now correct and the TODO should be removed. I will fix this. Good catch.

Reply:
Actually this code (and the TODO) is already gone in the latest revision.

Reply:
Sorry, it's gone in the next PR (https://github.com/appleseedhq/appleseed/pull/2756) which significantly improves the chart widget.